presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkNativeTaskRdd.java

```diff
@@ -0,0 +1,185 @@
+            checkArgument(shuffleRdd instanceof ShuffledRDD, format("ShuffledRdd is required but got: %s", shuffleRdd.getClass().getName()));
+
+            ShuffleHandle handle = ((ShuffleDependency<?, ?, ?>) shuffleRdd.dependencies().head()).shuffleHandle();
+            shuffleReadDescriptors.put(
+                    shuffleInputFragmentIds.get(i),
+                    new PrestoSparkShuffleReadDescriptor(
+                            partition,
+                            handle,
+                            shuffleRdd.getNumPartitions(),
+                            getBlockIds(((ShuffledRDDPartition) partition), handle),
+                            getPartitionSize(((ShuffledRDDPartition) partition), handle)));
+        }
+        return shuffleReadDescriptors.build();
+    }
+
+    private Optional<PrestoSparkShuffleWriteDescriptor> getShuffleWriteDescriptor(Partition split)
+    {
+        // Get shuffle information from Spark shuffle manager for shuffle write
+        checkArgument(
+                SparkEnv.get().shuffleManager() instanceof PrestoSparkNativeExecutionShuffleManager,
+                format("Native execution requires to use PrestoSparkNativeExecutionShuffleManager. But got: %s", SparkEnv.get().shuffleManager().getClass().getName()));
+        PrestoSparkNativeExecutionShuffleManager shuffleManager = (PrestoSparkNativeExecutionShuffleManager) SparkEnv.get().shuffleManager();
+        Optional<ShuffleHandle> shuffleHandle = shuffleManager.getShuffleHandle(split.index());
+
+        return shuffleHandle.map(handle -> new PrestoSparkShuffleWriteDescriptor(handle, shuffleManager.getNumOfPartitions(handle.shuffleId())));
+    }
+
+    private List<String> getBlockIds(ShuffledRDDPartition partition, ShuffleHandle shuffleHandle)
+    {
+        MapOutputTracker mapOutputTracker = SparkEnv.get().mapOutputTracker();
+        Collection<Tuple2<BlockManagerId, Seq<Tuple2<BlockId, Object>>>> mapSizes = asJavaCollection(mapOutputTracker.getMapSizesByExecutorId(
+                shuffleHandle.shuffleId(), partition.idx(), partition.idx() + 1));
+        return mapSizes.stream().map(item -> item._1.executorId()).collect(Collectors.toList());
+    }
+
+    private List<Long> getPartitionSize(ShuffledRDDPartition partition, ShuffleHandle shuffleHandle)
+    {
+        MapOutputTracker mapOutputTracker = SparkEnv.get().mapOutputTracker();
+        Collection<Tuple2<BlockManagerId, Seq<Tuple2<BlockId, Object>>>> mapSizes = asJavaCollection(mapOutputTracker.getMapSizesByExecutorId(
+                shuffleHandle.shuffleId(), partition.idx(), partition.idx() + 1));
+        //Each partition/BlockManagerId can contain multiple blocks (with BlockId), here sums up all the blocks from each BlockManagerId/Partition
+        return mapSizes.stream().map(item -> seqAsJavaList(item._2).stream().map(item2 -> ((Long) item2._2)).reduce(0L, Long::sum)).collect(Collectors.toList());
```

Comment:
This stream chaining looks very convoluted and harder to understand.

Reply:
Yeah it also took me some time to understand. What I can do is separating it to multiple lines for easier reading.